MainWindow/MainWindow.cpp

```diff
@@ -0,0 +1,119 @@
+#include <GlobalNames.h>
+#include <MainWindow/MainWindow.h>
+#include <MdiChild/MdiChild.h>
+#include "ui_mainwindow.h"
+
+MainWindow::MainWindow(QWidget *parent)
+    : QMainWindow(parent),
+      ui(new Ui::MainWindow),
+      mdiArea(new QMdiArea)
+{
+  ui->setupUi(this);
+  mdiArea->setHorizontalScrollBarPolicy(Qt::ScrollBarAsNeeded);
+  mdiArea->setVerticalScrollBarPolicy(Qt::ScrollBarAsNeeded);
+
+  setCentralWidget(mdiArea);
+  connect(mdiArea, &QMdiArea::subWindowActivated, this, &MainWindow::updateMenus);
+
+  createActions();
+  createStatusBar();
+  setWindowTitle(GLOBAL_STRINGS::PROGRAM_INFO::PROGRAM_NAME);
+}
+
+MainWindow::~MainWindow()
+{
+  delete ui;
+}
+void MainWindow::closeEvent(QCloseEvent *event)
+{
+  QWidget::closeEvent(event);
+  mdiArea->closeAllSubWindows();
+
+  if(mdiArea->currentSubWindow())
+  {
+    event->ignore();
+  } else
+  {
+    event->accept();
+  }
+}
+void MainWindow::newFile()
+{
+  MdiChild *child = createMdiChild();
+  child->newFile();
+  child->show();
+  statusBar()->showMessage("File created", 2000);
+}
+void MainWindow::open()
+{
+  const QString fileName = QFileDialog::getOpenFileName(this);
+  if (!fileName.isEmpty())
+  {
+    openFile(fileName);
+  }
+}
+bool MainWindow::openFile(const QString &fileName)
+{
+  if (QMdiSubWindow *existing = findMdiChild(fileName))
```

Comment:
Wrong assumption that you can't open twice the same file.

Let's say you have opened one file, then some modifications on image were done. Next you try to open the same file using menu. The program should open the same file but in the new window (two copies of the same files, but both of them may be modified independently).

Reply:
Ok, if this is our assumption for this behavior.